arc.mm

```diff
@@ -316,7 +326,10 @@ static inline id retain(id obj, BOOL isWeak)
 	if (isPersistentObject(obj)) { return obj; }
 	Class cls = obj->isa;
 	if ((Class)&_NSConcreteMallocBlock == cls ||
-	    (Class)&_NSConcreteStackBlock == cls)
+	    (Class)&_NSConcreteStackBlock == cls ||
+		(Class)&_NSConcreteAutoBlock == cls ||
+		(Class)&_NSConcreteFinalizingBlock == cls
+		) 
```

Comment:
Small complication here is that  `objc_create_block_classes_as_subclasses_of()` only gets called when a Foundation library sets it up explicitly with its root class, so I had to pre-initialise just the info field when the runtime loads.

btw: I'm assuming that stack blocks are always promoted to the heap before they are assigned to weak references. Is that an accurate assumption?` (we need a couple of checks for `_NSConcreteStackBlock` otherwise).

Reply:
I’m not sure. I vaguely remember being able to take a weak reference to a block and have it go away when the block scope ended, but I’m not completely sure.

Reply:
So I've looked into this a bit more and did some experimentation and I think it's safe the way it is right now. The point I was worried about was what happens in `objc_storeWeak()`, where we don't do anything to tie the lifetime of a stack block to that of the weak reference. But since this is all happening under ARC, the compiler will always emit a call to `objc_retainBlock()` if there is any chance that the block will escape the current scope – at least that's my reading of `CodeGenFunction::EmitARCRetainBlock()` in LLVM.